mlflow/pipelines/steps/automl/flaml.py

```diff
@@ -0,0 +1,124 @@
+    )
+
+
+def _create_custom_metric_flaml(metric_name: str, coeff: int, custom_func: callable) -> callable:
+    def add_suffix(metrics: Dict[str, float], suffix: str) -> Dict[str, float]:
+        return {f"{k}_{suffix}": v for k, v in metrics.items()}
+
+    def calc_metric(X, y, estimator) -> Dict[str, float]:
+        y_pred = estimator.predict(X)
+        builtin_metrics = _get_regressor_metrics(y, y_pred, sample_weights=None)
+        res_df = pd.DataFrame()
+        res_df["prediction"] = y_pred
+        res_df["target"] = y.values
+        return custom_func(res_df, builtin_metrics)
+
+    # pylint: disable=keyword-arg-before-vararg
+    # pylint: disable=unused-argument
+    def custom_metric(
+        X_val,
+        y_val,
+        estimator,
+        labels,
+        X_train,
+        y_train,
+        weight_val=None,
+        weight_train=None,
+        *args,
+    ):
+        val_metrics = calc_metric(X_val, y_val, estimator)
+        if metric_name not in val_metrics:
+            raise MlflowException(
+                f"User-defined function has not calculated expected primary metric {metric_name}.\n"
+                f"The function has returned the following metrics: {val_metrics}"
+            )
+        main_metric = coeff * val_metrics[metric_name]
+        val_metrics = add_suffix(val_metrics, "val")
+        train_metrics = add_suffix(calc_metric(X_train, y_train, estimator), "train")
+        return main_metric, {**val_metrics, **train_metrics}
+
+    return custom_metric
+
+
+def _create_model_automl(
+    X,
+    y,
+    step_config: Dict[str, Any],
+    pipeline_root: str,
+    evaluation_metrics: Dict[str, PipelineMetric],
+    primary_metric: str,
+) -> Tuple[BaseEstimator, Dict[str, Any]]:
+    try:
+        from flaml import AutoML
+    except ImportError:
+        raise MlflowException("Please install FLAML to use AutoML!")
+
+    try:
+        if primary_metric in MLFLOW_TO_FLAML_METRICS and primary_metric in evaluation_metrics:
+            metric = MLFLOW_TO_FLAML_METRICS[primary_metric]
+        elif primary_metric in evaluation_metrics:
+            metric = _create_custom_metric_flaml(
+                primary_metric,
+                -1 if evaluation_metrics[primary_metric].greater_is_better else 1,
+                _load_one_custom_metric_function(pipeline_root, evaluation_metrics[primary_metric]),
+            )
+        else:
+            _logger.warning(
+                f"There is no FLAML alternative or custom metric for {primary_metric} metric.\n"
+                f"Using 'auto' metric instead."
+            )
+            metric = "auto"
+
+        automl_settings = step_config.get("flaml_params", {})
+        automl_settings["time_budget"] = step_config.get(
+            "time_budget_secs", AUTOML_DEFAULT_TIME_BUDGET
+        )
+        automl_settings["metric"] = metric
+        automl_settings["task"] = "regression"
+
+        mlflow.autolog(disable=True)
```

Comment:
Why do we have to disable autologging here? In https://github.com/microsoft/FLAML/pull/747, we fixed the mlflow-FLAML integration so that the logging is in the form we want.

Reply:
If I comment out this line, I get a weird exception in our auto-logging code. Btw, it looks like an MLflow bug: 
```
Traceback (most recent call last):
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/pipelines/steps/automl/flaml.py", line 114, in _create_model_automl
    automl.fit(X, y, **automl_settings)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/flaml/automl.py", line 2898, in fit
    self._search()
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/flaml/automl.py", line 3462, in _search
    self._search_sequential()
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/flaml/automl.py", line 3280, in _search_sequential
    analysis = tune.run(
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/flaml/tune/tune.py", line 514, in run
    result = evaluation_function(trial_to_run.config)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/flaml/automl.py", line 350, in _compute_with_config_base
    ) = compute_estimator(
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/flaml/ml.py", line 610, in compute_estimator
    val_loss, metric_for_logging, train_time, pred_time = evaluate_model_CV(
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/flaml/ml.py", line 523, in evaluate_model_CV
    val_loss_i, metric_i, train_time_i, pred_time_i = get_val_loss(
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/flaml/ml.py", line 412, in get_val_loss
    estimator.fit(X_train, y_train, budget, **fit_kwargs)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/flaml/model.py", line 1133, in fit
    self._fit(
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/flaml/model.py", line 197, in _fit
    model.fit(X_train, y_train, **kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 555, in safe_patch_function
    patch_function(call_original, *args, **kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 254, in patch_with_managed_run
    result = patch_function(original, *args, **kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/sklearn/__init__.py", line 1552, in patched_fit
    result = fit_impl(original, self, *args, **kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/sklearn/__init__.py", line 1299, in fit_mlflow_xgboost_and_lightgbm
    fit_output = original(self, *args, **kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 536, in call_original
    return call_original_fn_with_event_logging(_original_fn, og_args, og_kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 471, in call_original_fn_with_event_logging
    original_fn_result = original_fn(*og_args, **og_kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 533, in _original_fn
    original_result = original(*_og_args, **_og_kwargs)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/lightgbm/sklearn.py", line 895, in fit
    super().fit(X, y, sample_weight=sample_weight, init_score=init_score,
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 555, in safe_patch_function
    patch_function(call_original, *args, **kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 254, in patch_with_managed_run
    result = patch_function(original, *args, **kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/sklearn/__init__.py", line 1559, in patched_fit
    return original(self, *args, **kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 536, in call_original
    return call_original_fn_with_event_logging(_original_fn, og_args, og_kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 471, in call_original_fn_with_event_logging
    original_fn_result = original_fn(*og_args, **og_kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 533, in _original_fn
    original_result = original(*_og_args, **_og_kwargs)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/mlflow-dev-env/lib/python3.8/site-packages/lightgbm/sklearn.py", line 748, in fit
    self._Booster = train(
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 555, in safe_patch_function
    patch_function(call_original, *args, **kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/safety.py", line 254, in patch_with_managed_run
    result = patch_function(original, *args, **kwargs)
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/lightgbm.py", line 665, in train
    param_logging_operations.await_completion()
  File "/Users/michael.shtelma/Development/Projects/mlflow_msh/mlflow/utils/autologging_utils/client.py", line 69, in await_completion
    raise MlflowException(
mlflow.exceptions.MlflowException: The following failures occurred while performing one or more logging operations: [MlflowException('Failed to perform one or more operations on the run with ID 91aa473ed5cd417184e4de742889c273. Failed operations: [MlflowException(\'Changing param values is not allowed. Params were already logged=\\\'[{\\\'key\\\': \\\'boosting_type\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'gbdt\\\'}, {\\\'key\\\': \\\'max_depth\\\', \\\'old_value\\\': \\\'0\\\', \\\'new_value\\\': \\\'-1\\\'}, {\\\'key\\\': \\\'min_child_samples\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'20\\\'}, {\\\'key\\\': \\\'min_child_weight\\\', \\\'old_value\\\': \\\'0.9999999999999993\\\', \\\'new_value\\\': \\\'0.001\\\'}, {\\\'key\\\': \\\'min_split_gain\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'0.0\\\'}, {\\\'key\\\': \\\'num_leaves\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'4\\\'}, {\\\'key\\\': \\\'subsample_for_bin\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'200000\\\'}, {\\\'key\\\': \\\'subsample_freq\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'0\\\'}, {\\\'key\\\': \\\'max_bin\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'255\\\'}, {\\\'key\\\': \\\'verbose\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'-1\\\'}, {\\\'key\\\': \\\'objective\\\', \\\'old_value\\\': \\\'reg:squarederror\\\', \\\'new_value\\\': \\\'regression\\\'}, {\\\'key\\\': \\\'metric\\\', \\\'old_value\\\': None, \\\'new_value\\\': "[\\\'regression\\\']"}, {\\\'key\\\': \\\'feature_name\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'auto\\\'}, {\\\'key\\\': \\\'categorical_feature\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'auto\\\'}, {\\\'key\\\': \\\'evals_result\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'None\\\'}, {\\\'key\\\': \\\'verbose_eval\\\', \\\'old_value\\\': \\\'True\\\', \\\'new_value\\\': \\\'warn\\\'}, {\\\'key\\\': \\\'keep_training_booster\\\', \\\'old_value\\\': None, \\\'new_value\\\': \\\'False\\\'}]\\\' for run ID=\\\'91aa473ed5cd417184e4de742889c273\\\'.\')]')]
```

Reply:
https://databricks.atlassian.net/browse/ML-26033